Added a primer on CMake to docs and tweak source file search (#272)
enzo-project/enzo-e: doc/source/devel/cmake_primer.rst

***************
Primer on CMake
***************

There is a lot of excellent documentation for CMake available online.
However, some of that documentation can seem overwhelming.

This primer was written to provide some information at a level between the :ref:`getting_started` tutorial and that documentation.

As an aside, when searching for online CMake documentation and guides, you should generally make sure that the guide was written for version 3.0 or later (sometimes called "Modern CMake").

================================
History of Enzo-E's build system
================================

> Historically, Enzo-E made use of the `SCons <https://scons.org/>`_ build system. Howver, inn 2022 Enzo-E migrated to the CMake build system.

Reviewer: ```suggestion
Historically, Enzo-E made use of the `SCons <https://scons.org/>`_ build system. However, in 2022 Enzo-E migrated to the CMake build system.
```

> Historically, Enzo-E made use of the `SCons <https://scons.org/>`_ build system. Howver, inn 2022 Enzo-E migrated to the CMake build system.

Author: ```suggestion
Historically, Enzo-E made use of the `SCons <https://scons.org/>`_ build system. However, inn 2022 Enzo-E migrated to the CMake build system.
```

While there are a number of viable build systems, there has been a general convergence among C++ developers towards using CMake (this is evidenced by large projects like Boost and Charm++).
Given the very large userbase of CMake, bugs are found and fixed relatively quickly.
Moreover, CMake makes the process of locating and linking against external dependencies relatively easy (especially if they are popular libraries OR were themselves built with CMake).

In contrast, SCons is relatively less popular.
This means that certain areas of the codebase receive less attention (e.g. less widely used compiler toolchains), which can lead to some issues that are more difficult to work around.
Additionally, support staff at clusters are less likely to be familiar with SCons (so they are generally less equipped to help support users)

.. _how_a_cmake_build_works:

============================
How Does a CMake Build Work?
============================

> For people mostly familar with build systems like makefiles, CMake may seem a little unintuitive.

Reviewer: ```suggestion
For people mostly familiar with build systems like makefiles, CMake may seem a little unintuitive.
```

..  code-block:: bash

  # these commands are typically executed from the root directory of the
  # Enzo-E repository
  mkdir build
  cd build

2. The next step is to actually invoke CMake.
   This step involves the configuration of your build (more details about various configuration options for Enzo-E are provided in :ref:`how-to_configure_build`).
   It's important to understand that CMake does not generally compile any code [#f1]_.
   Instead, this step generates build-files (stored within the build-directory) that are natively understood by the user's platform.
   For most Enzo-E user/developers (on unix-like systems), CMake will generate makefiles or (optionally) ninja files.
   When you invoke cmake on the command-line you typically include variables related to the configuration and the final argument specifies the path to the root of the Enzo-E directory.
   If you placed your build file at the root of your Enzo-E directory, this step might look something like:

..  code-block:: bash

>   cmake -D-DCHARM_ROOT=<PATH/TO/charm/build-dir> \

Reviewer: ```suggestion
  cmake -DCHARM_ROOT=<PATH/TO/charm/build-dir> \
```

If there are any differences in the recovered set of files, CMake reconstructs the build system.
The CMake developers generally view this unfavorably because:

 * globbing expressions can be expensive (this may be less of an issue on Linux compared to windows)

 * "[t]he ``CONFIGURE_DEPENDS`` flag may not work reliably on all generators"

For that reason, we may want to reconsider our approach in the future.

.. note:
   In the interval of time between Enzo-E's transition to using CMake and the patch introducing this documentation, the command to locate source files with globbing expressions did not include the ``CONFIGURE_DEPENDS`` Flag.
   For reasons described above, steps generally needed to be taken to force the regeneration of the build system (e.g. deleting the build and starting again).
   This should now be less of an issue.

What is Ninja?
~~~~~~~~~~~~~~

`Ninja <https://ninja-build.org/>`_ is an alternative build-system to something like Make.
Ninja was designed to be a smaller, less-feature-rich alternative to Make that is intended to be used with tools (like CMake) that generate its input files.

> These design goals can facillitate faster build-times in large projects.

Reviewer: ```suggestion
These design goals can facilitate faster build-times in large projects.
```

> To use ``Ninja``, you need to make sure it is installed on your machine and you need to specify ``-GNinja`` as one of the configuration arguments in the second step described in :ref:`how_a_cmake_build_works`.

Reviewer: ```suggestion
To use ``Ninja``, you need to make sure it is installed on your machine and you need to specify ``-G Ninja`` as one of the configuration arguments in the second step described in :ref:`how_a_cmake_build_works`.
```
